### DISCUSSION TITLE
About Different Use Cases for NativeMenu

### DISCUSSION THREAD

**Discussion by gebodal** [AvaloniaUI/Avalonia#16589] [Q&A]
Hello all,

Slightly as a follow-up to [this question](https://github.com/AvaloniaUI/Avalonia/discussions/15560), I was hoping someone could elaborate on the different uses of `NativeMenu` in Avalonia apps as of 10.x/11 - as it seems things have moved on a bit since some of the answers I could find were written.

What exactly are the rules for including a `NativeMenu` in an application? (I'm interested specifically in desktop apps, and I'm assuming that's pretty much the limit of the `NativeMenu`...?)

More specifically:
1. I've seen references to it in both `App.axaml` and `MainWindow.axaml` (or equivalents), included under `<NativeMenu.Menu>` as a child of `<Application>` and `<Window>`, respectively. Are these still appropriate locations?
2. What is the difference between putting the `NativeMenu` in the `<Application>` versus the `<Window>`?
3. Can different windows have different `NativeMenu` objects in their AXAML? How does this relate to the main "application" drop-down menu seen on MacOS (i.e. the one that is usually boldfaced containing the application name)?
4. I've also seen `NativeMenu` used under `<TrayIcon.Menu>` in `<Application>`. What operating systems does this work with? Can this also be used with `<Window>`? Will this cause the menu to only appear as a context menu of the tray icon, or will it also appear in the system bar (as for MacOS)?
5. For the main system bar (i.e. MacOS top bar), how do we determine if it is being shown to the user? The use case I'm imagining here is: I have a `Menu` with the standard "File"/"Edit"/"View" etc in my `MainWindow.axaml`, the structure of which is mimicked in a `NativeMenu` (placed _somewhere_, as above). Only one of these menus should be visible on the application, so what do I bind the `IsEnabled`/`IsVisible` properties of the `Window`'s `Menu` to? I'm assuming it's somehow related to `IsNativeMenuExportedProperty`?
6. As a follow-up, are there any good practices for having a `Menu`/`NativeMenu` mimic each other in terms of content, so they don't have to be maintained separately? I could imagine writing a script to convert a `Menu` into a `NativeMenu` (throwing away icons and the like), but is this a good approach?
7. (Following follow-up) Do `NativeMenuItem` objects support `ItemsSource` equivalents for dynamically adjusting content?
8. Are there any other standard locations for `<NativeMenu>`?
9. Are there any other standard uses for `<NativeMenu>`?

([The docs](https://docs.avaloniaui.net/docs/reference/controls/nativemenu) could possibly do with improving in this area, as they don't discuss what kinds of controls NativeMenu can be included with, other than that it can be placed inside a tray icon.)

Any wisdom on the matter is much appreciated.

**Comment by nesteruk** [comment]
NativeMenu is one of the most cryptic things in avalonia
